File: Sources/MDifyCore/Support/ConvertibleFilePolicy.swift
import Foundation

public struct ConvertibleFilePolicy: Sendable {
    public static let `default` = ConvertibleFilePolicy()

    public let supportedExtensions: Set<String>

    public init(supportedExtensions: Set<String> = ConvertibleFilePolicy.defaultExtensions) {
        self.supportedExtensions = Set(supportedExtensions.map { $0.lowercased() })
    }

    public init(workerKind: WorkerKind) {
        switch workerKind {
        case .lite:
            self.init(supportedExtensions: Self.liteExtensions)
        case .ocr:
            self.init(supportedExtensions: Self.ocrExtensions)
        }
    }

    public func isConvertibleFile(_ url: URL) -> Bool {
        let ext = url.pathExtension.lowercased()
        return !ext.isEmpty && supportedExtensions.contains(ext)
    }

    public static let liteExtensions: Set<String> = [
        "pdf",
        "docx", "pptx", "xlsx", "xls",
        "html", "htm", "csv", "json", "xml", "txt", "md",
        "epub", "zip",
        "jpg", "jpeg", "png", "bmp", "tif", "tiff", "webp"
    ]

    public static let ocrExtensions: Set<String> = liteExtensions

    public static let defaultExtensions: Set<String> = ocrExtensions
}

File: windows/MDify.Windows/Support/ConvertibleFilePolicy.cs
using System.IO;
using MDify.Windows.Models;

namespace MDify.Windows.Support;

public sealed class ConvertibleFilePolicy
{
    public static readonly IReadOnlySet<string> LiteExtensions = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
    {
        ".pdf",
        ".docx",
        ".pptx",
        ".xlsx",
        ".xls",
        ".html",
        ".htm",
        ".csv",
        ".json",
        ".xml",
        ".txt",
        ".md",
        ".zip",
        ".epub"
    };

    public static readonly IReadOnlySet<string> OcrImageExtensions = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
    {
        ".jpg",
        ".jpeg",
        ".png",
        ".tif",
        ".tiff",
        ".webp",
        ".bmp"
    };

    public static readonly IReadOnlySet<string> OcrExtensions = new HashSet<string>(
        LiteExtensions.Concat(OcrImageExtensions),
        StringComparer.OrdinalIgnoreCase);

    public static ConvertibleFilePolicy Default { get; } = new(WorkerKind.Ocr);

    private readonly IReadOnlySet<string> _supportedExtensions;

    public ConvertibleFilePolicy(WorkerKind workerKind = WorkerKind.Ocr)
        : this(workerKind == WorkerKind.Lite ? LiteExtensions : OcrExtensions)
    {
    }

    public ConvertibleFilePolicy(IEnumerable<string> supportedExtensions)
    {
        _supportedExtensions = new HashSet<string>(
            supportedExtensions.Select(NormalizeExtension),
            StringComparer.OrdinalIgnoreCase);
    }

    public bool IsConvertibleFile(string path)
    {
        var extension = Path.GetExtension(path);
        return !string.IsNullOrWhiteSpace(extension) && _supportedExtensions.Contains(extension);
    }

    public bool IsLiteSupported(string path) => LiteExtensions.Contains(Path.GetExtension(path));

    public bool IsOcrSupported(string path) => OcrExtensions.Contains(Path.GetExtension(path));

    public bool IsImage(string path) => OcrImageExtensions.Contains(Path.GetExtension(path));

    public bool IsPdf(string path) => string.Equals(Path.GetExtension(path), ".pdf", StringComparison.OrdinalIgnoreCase);

    private static string NormalizeExtension(string extension)
    {
        if (string.IsNullOrWhiteSpace(extension))
        {
            return string.Empty;
        }

        return extension.StartsWith(".", StringComparison.Ordinal) ? extension : $".{extension}";
    }
}
